mbedtls/src/ssl/context.rs

```diff
@@ -172,35 +172,105 @@ impl<T> Context<T> {
     }
 }
 
+/// # Safety
+/// `io` must live as long as `ctx` or the next time bio is set/cleared.
+unsafe fn set_bio_raw<IoType, T: IoCallbackUnsafe<IoType>>(ctx: *mut ssl_context, io: &mut T) {
+    ssl_set_bio(
+        ctx,
+        io as *mut T as *mut c_void,
+        Some(T::call_send),
+        Some(T::call_recv),
+        None,
+    );
+}
+
+/// This function provides a way to apply async context to bio before running 
+/// any logic.
+/// Note: `bio` is a concept in common TLS implementation which refers to basic IO.
+/// openssl and mbedtls both use this concept.
+/// Ref: https://stackoverflow.com/questions/51672133/what-are-openssl-bios-how-do-they-work-how-are-bios-used-in-openssl
+#[cfg(all(feature = "std", feature = "async"))]
+impl<T> Context<T>  {
+    pub(super) fn with_bio_async<'cx, R, IoType>(&mut self, cx: &mut std::task::Context<'cx>, f: impl FnOnce(&mut Self) -> R) -> Option<R> where for<'c> (&'c mut std::task::Context<'cx>, &'c mut T): IoCallbackUnsafe<IoType> {
+        let ret;
+
+        struct BioGuard<'a, T> {
+            context: &'a mut Context<T>,
+        }
+        
+        impl<'a, T> Drop for BioGuard<'a, T> {
+            fn drop(&mut self) {
+                self.context.clear_bio();
+            }
+        }
+        // SAFETY: In the call to `set_bio_raw`, `user_data` must live as long
+        // as `ctx`, or until the bio is cleared from `ctx`. The bio is cleared
+        // at the end of this block ensured by the drop guard: [`BioGuard`]
+        unsafe {
+            // Points to `self.inner`, so safe to borrow at the same time as `self.io`
+            let ctx = self.into();
+            let mut user_data = (cx, &mut**self.io.as_mut()?);
+            set_bio_raw(ctx, &mut user_data);
+
+            let guard = BioGuard { context: self };
+
+            ret = f(guard.context);
+        }
+
+        Some(ret)
+    }
+
+    // This function is created to handle the ood behavior of `mbedtls_ssl_write()` only for TLS use cases
+    // Please check this https://github.com/Mbed-TLS/mbedtls/issues/4183 to learn more about how `mbedtls_ssl_write()` works in c-mbedtls 2.28
+    // This function ultimately ensure the semitics:
+    // Returned value `Ok(n)` always means n bytes of data has been sent into c-mbedtls's buffer (some of them might be sent out through underlying IO)
```

Comment:
What does "only for TLS use cases" mean? That seems redundant, if you have an ssl_context you're doing TLS.

Typos:
* ood -> odd
* semitics -> semantics

Reply:
Will update it.
What I mean here is more about this function might not work with DTLS